Assets/Mirror/Runtime/SyncDictionary.cs

```diff
@@ -0,0 +1,322 @@
+                        key = DeserializeKey(reader);
+                        item = DeserializeItem(reader);
+                        if (apply)
+                        {
+                            m_Objects.Remove(key);
+                        }
+                        break;
+
+                    case Operation.OP_SET:
+                    case Operation.OP_DIRTY:
+                        key = DeserializeKey(reader);
+                        item = DeserializeItem(reader);
+                        if (apply)
+                        {
+                            m_Objects[key] = item;
+                        }
+                        break;
+                }
+
+                if (apply)
+                {
+                    Callback?.Invoke(operation, key, item);
+                }
+                // we just skipped this change
+                else
+                {
+                    changesAhead--;
+                }
+            }
+        }
+
+
+        public void Clear()
+        {
+            m_Objects.Clear();
+            AddOperation(Operation.OP_CLEAR, default, default);
+        }
+
+        public bool ContainsKey(B key) => m_Objects.ContainsKey(key);
+
+        public bool Remove(B key)
+        {
+            if (m_Objects.ContainsKey(key))
+            {
+                T item = m_Objects[key];
+                bool result = m_Objects.Remove(key);
+                if (result)
+                {
+                    AddOperation(Operation.OP_REMOVE, key, item);
+                }
+                return result;
+            }
+            else
+            {
+                return false;
+            }
+        }
+
+        public void Dirty(B index)
+        {
+            AddOperation(Operation.OP_DIRTY, index, m_Objects[index]);
+        }
+
+        public T this[B i]
+        {
+            get => m_Objects[i];
+            set
+            {
+                bool existing = TryGetValue(i, out var val);
+                if (existing)
+                {
+                    m_Objects[i] = value;
+                    AddOperation(Operation.OP_SET, i, value);
+                }
+                else
+                {
+                    m_Objects[i] = value;
+                    AddOperation(Operation.OP_ADD, i, value);
+                }
+            }
+        }
+
+        public bool TryGetValue(B key, out T value) => m_Objects.TryGetValue(key, out value);
+
+        public void Add(B key, T value)
+        {
+            m_Objects.Add(key, value);
+            AddOperation(Operation.OP_ADD, key, value);
+        }
+
+        public void Add(KeyValuePair<B, T> item)
+        {
+            m_Objects.Add(item.Key, item.Value);
+            AddOperation(Operation.OP_ADD, item.Key, item.Value);
+        }
+
+        public bool Contains(KeyValuePair<B, T> item)
+        {
+            return m_Objects.ContainsKey(item.Key) && m_Objects.ContainsValue(item.Value);
```

Comment:
The performance of ContainsValue is bad, its not a good access pattern since it has to go through **all** the dictionary values. Besides that, its plain wrong, as some other key could have this as a value. You should use TryGetValue with the item Key, then check that the value matches.

Reply:
@Katori please write a unit test that covers what @rodolphito described